### PR TITLE
tests: detect and report root-owned files in /home

### PR DESCRIPTION
**Pull request by zyga**
This patch adds a tool that look for root-owned files in /home. This
was found to break tests, particularly the pulseaudio test, which relies
on ability to create write to ~/.config as the non-root, test user.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

**Comment by zyga** [comment]
This has picked up a minor bug in how our core16 and core18 preparation works. We seem to be copying the /home/ubuntu directory over from classic. I'll send a patch to address that shortly.

File: tests/lib/bin/invariant-tool [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#!/bin/sh
+if [ $# -eq 0 ]; then
+	echo "usage: invariant-tool check [INVARIANT...]"
```
Comment (stolowski) [inline tests/lib/bin/invariant-tool:3]: Nitpick, I think this should also show available invariants, it's not clear what the allowed arguments are.

File: tests/lib/bin/invariant-tool [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+			;;
+		*)
+			break
+			;;
+	esac
+done
+
+check_root_files_in_home() {
+	n="$1" # invariant name
+	find /home -mindepth 1 -user root -a ! -path "$GOHOME" -a ! -path "$GOHOME/*" 2>/dev/null >"/tmp/invariant-tool.$n"
```
Comment (stolowski) [inline tests/lib/bin/invariant-tool:30]: I wonder if this could be simplified by just doing find over the only test user we have in the test (or do we have more?) and gettting rid of `-path ...` exclusions.

File: tests/lib/prepare-restore.sh [hunk 3]
```diff
@@ -530,6 +530,9 @@ prepare_suite_each() {
             ausearch -i -m AVC --checkpoint "$RUNTIME_STATE_PATH/audit-stamp" || true
             ;;
     esac
+
+    # Check for invariants late, in order to detect any bugs in the code above.
```
Comment (stolowski) [inline tests/lib/prepare-restore.sh:534]: Nice.

File: tests/lib/prepare.sh [hunk 4]
```diff
@@ -881,12 +881,24 @@ EOF
...
         setup_reflash_magic
         REBOOT
     fi
 
+    # XXX: we've just rebooted and cloud-init created a broken account for the
+    # "ubuntu" user.  The account has a home directory /home/ubuntu but is
+    # actually owned by root:root.
+    # Open questions:
+    # - Why does this happen, is it cloud-init?
+    # - Why is the ownership wrong?
```
Comment (stolowski) [inline tests/lib/prepare.sh:896]: Can this be a separate PR?

File: tests/main/invariant-tool/task.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+summary: invariant-tool is detecting problems
```
Comment (stolowski) [inline tests/main/invariant-tool/task.yaml:1]: I think we should introduce a new category besides main/ for tests of helper tools (session-tool would go there too).

File: tests/lib/bin/invariant-tool (hunk 2 above)
Comment (zyga) [reply tests/lib/bin/invariant-tool:30]: Yes but:
 - we have a special test that adds a test-zsh user, that would not catch it
 - we have the magic /home/ubuntu that picked up a what looks like, a real bug with cloud-init

File: tests/main/invariant-tool/task.yaml (hunk 5 above)
Comment (zyga) [reply tests/main/invariant-tool/task.yaml:1]: Yeah, I was thinking about it on Friday. It would be extra useful IF it did not build snapd and the rest of the stack, and would just allow us to do the quick testing of the helper tools themselves.

File: tests/lib/prepare.sh (hunk 4 above)
Comment (zyga) [reply tests/lib/prepare.sh:896]: I you feel strongly about it, sure. Is there a particular worry you want to address here?

Comment (stolowski) [reply tests/lib/prepare.sh:896]: My worry is that we don't know where does it come from and this hides the problem. So I'm not convinced we should land this change (while the rest of this PR is uncontroversial).

File: tests/lib/bin/invariant-tool (hunk 2 above)
Comment (stolowski) [reply tests/lib/bin/invariant-tool:30]: Yes, you're right, I didn't know of extra users.

File: tests/lib/prepare.sh (hunk 4 above)
Comment (zyga) [reply tests/lib/prepare.sh:896]: We know this is coming from cloud-image. Beyond that it's something to debug

Comment (zyga) [reply tests/lib/prepare.sh]: I've dropped this now. We just ignore the /home/ubuntu ownership.

File: tests/lib/bin/invariant-tool (hunk 1 above)
Comment (zyga) [reply tests/lib/bin/invariant-tool:4]: Done.

File: tests/lib/bin/invariant-tool [hunk 6]
```diff
@@ -0,0 +1,93 @@
+#!/bin/sh
+
+show_help() {
+	echo "usage: invariant-tool check [INVARIANT...]"
+	echo
+	echo "Supported invariants:"
+	echo "    root-files-in-home: most of /home/* does not contain root-owned files"
```
Comment (stolowski) [inline tests/lib/bin/invariant-tool:7]: Thank you.

File: tests/main/invariant-tool/task.yaml (hunk 5 above)
Comment (stolowski) [reply tests/main/invariant-tool/task.yaml:1]: Something to possibly revisit in near future, not for this PR; we're growing tests for tooling, unrelated to snapd functionality, would be nice to separate them.

File: tests/lib/bin/invariant-tool [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	n="$1" # invariant name
+	# This find expression looks for root owned file in /home with the following exceptions:
+	# - The /home/ubuntu directory is root owned but is not used by the tests so we ignore it
+	# - The /home/gopath and everything inside it comes from spread.yaml and is too inconvenient to change
+	# As a note, the working theory for the origin of /home/ubuntu is cloud-init and data from GCE.
+	find /home -mindepth 1 -user root -a ! -path "/home/ubuntu" -a ! -path "$GOHOME" -a ! -path "$GOHOME/*" 2>/dev/null >"/tmp/invariant-tool.$n"
+	if [ -s "/tmp/invariant-tool.$n" ]; then
+		echo "invariant-tool: the following files should not be owned by root" >&2
+		cat "/tmp/invariant-tool.$n" >&2
+		return 1
```
Comment (stolowski) [inline tests/lib/bin/invariant-tool:50]: Thanks!

File: tests/main/invariant-tool/task.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    # The root-files-in-home invariant detects files owned by root anywhere in /home/
+    invariant-tool check root-files-in-home | MATCH 'invariant-tool: root-files-in-home ok'
+    touch /home/test/invariant-canary
+    echo 'rm -f /home/test/invariant-canary' >> defer.sh
+    chown root /home/test/invariant-canary
+    invariant-tool check root-files-in-home 2>&1 | MATCH 'invariant-tool: root-files-in-home not-ok'
+    invariant-tool check root-files-in-home 2>&1 | MATCH 'invariant-tool: the following files should not be owned by root'
+    invariant-tool check root-files-in-home 2>&1 | MATCH '/home/test/invariant-canary'
+    invariant-tool check root-files-in-home 2>&1 | MATCH 'invariant-tool: system is corrupted'
```
Comment (anonymouse64) [inline tests/main/invariant-tool/task.yaml:31]: nice way to handle multi-line output matching!

/me will steal this trick for my lxd PR

File: tests/main/invariant-tool/task.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    not invariant-tool check root-files-in-home 2>/dev/null
+
+    # Chown is the solution.
+    chown test /home/test/invariant-canary
+    invariant-tool check root-files-in-home | MATCH 'invariant-tool: root-files-in-home ok'
+
+    # The root-files-in-home invariant ignores /home/gopath which comes from spread.yaml
+    test -d /home/gopath
+    test "$(stat -c %u /home/gopath)" -eq 0
```
Comment (anonymouse64) [inline tests/main/invariant-tool/task.yaml:40]: thanks for testing this too so when we stop doing this we will remember to update the invariant check too

File: tests/main/invariant-tool/task.yaml (hunk 8 above)
Comment (zyga) [reply tests/main/invariant-tool/task.yaml:31]: It's not quite perfect but it tends to work in simple cases. I haven't found a way to make the many-line ordering test nice.
